mysql & mariadb: Move bind address 127.0.0.1 to my.cnf

npinchot opened the pull request:
#### Summary
Move the 127.0.0.1 bind address into a default `my.cnf` that is created during post install instead of having it as a param in the plist file.

#### Reason
Trying to help people (myself included) not have to spend a lot of time troubleshooting why external hosts can't connect.

#### Misc
Reference issue #12744.

I've updated both MySQL & MariaDB in this same pull request - please let me know if you prefer separate pull requests for each.

MikeMcQuaid commented on Formula/mariadb.rb:
```diff
@@ -151,6 +151,9 @@ def caveats; <<-EOS.undent
     To connect:
         mysql -uroot
+
+    Using brew services will place a plist file at #{ENV["HOME"]}/Library/LaunchAgents/#{plist_name}
```
I don't think this is the place for `brew services` documentation. It could be added to `--help` or the `README` in that repository instead.

npinchot replied:
Sure but it's not very helpful to say go edit a plist file without noting where that plist file is or even how to find it. I see that running `brew services -v start mysql` tells you it's placing a plist file and where it's placed. This was not obvious to me to even try because I didn't know the plist file was overriding the bind address. I think adding some info to the caveat is the quickest/simplest solve to this problem and will save people a lot of time troubleshooting.

How about replacing this instead?
```
To connect run:
    mysql -uroot

The plist file for this service overrides the bind address to 127.0.0.1.
Edit this file if you need to change the server's bind address.
Enable verbose mode when starting via brew services for more detail on the plist file.
```

npinchot replied:
If you don't like the 3rd line about enabling verbose mode for brew services it could be ditched. The first 2 lines give somebody a fighting chance of knowing what to try and look for to change the bind address and that it's even a thing that is being overridden.

MikeMcQuaid replied:
Problem is if you change the plist it'll be overwritten by `brew services` in future. I think the `my.cnf` approach is probably the best bet.

npinchot replied:
Okay sounds good and fair point. I can work on this, unless you would prefer to implement - let me know. Want me to close this PR for one with those changes or update this one?

MikeMcQuaid replied:
Update this one works for me, thanks.

MikeMcQuaid commented on Formula/mariadb.rb:
```diff
@@ -151,6 +151,9 @@ def caveats; <<-EOS.undent
         mysql -uroot
+
+    Using brew services will place a plist file at #{ENV["HOME"]}/Library/LaunchAgents/#{plist_name}
+    This plist file overrides the bind address to 127.0.0.1. Edit this plist file if you need to change the server's bind address.
```
Would this be more obvious if we had set it in the server configuration?

npinchot replied:
Do you mean putting it in `my.cnf` file in one of the locations? I think that would make it more obvious because most Google results I found about trying to change the MySQL bind address said to edit `my.cnf` in one of the proper locations. That doesn't work because the plist file overrides it. I figured it would be much simpler to just add some info to the caveat about how to fix the bind address if you need to.

MikeMcQuaid replied:
Yes. I think we could probably avoid the caveat entirely in that case.

npinchot commented:
@MikeMcQuaid I've implemented what I think is a decent first go at having the bind address in `my.cnf`. I've left it to `mysql.rb` at the moment only and if you like the results I'll update `mariadb.rb` as well. Please let me know your thoughts/feedback. I tried my best to follow standard formula conventions but if there is something I've missed that should be done differently please let me know.

MikeMcQuaid commented on Formula/mysql.rb:
```diff
@@ -123,13 +123,27 @@ def post_install
     end
+
+    # Create my.cnf that binds to 127.0.0.1 by default
+    unless (etc/"my.cnf").exist?
```
Use `(buildpath/"my.cnf").write <<-EOS.undent` then `etc.install "my.cnf"` which will automatically handle duplicate detection.

npinchot replied:
Thanks for the feedback on this! I really appreciate you sticking with me through this :)

To use `(buildpath/"my.cnf").write` looks like this logic needs to be moved to the `install` function? I found [the git](https://github.com/Homebrew/homebrew-core/blob/master/Formula/git.rb#L158) [and uwsgi](https://github.com/Homebrew/homebrew-core/blob/master/Formula/uwsgi.rb#L64) formulas do a similar thing but they do this in the `install` function. I tried leaving the logic in the `post_install` function but while there are no crashes, the file doesn't get copied. Additionally, `buildpath` is null when doing anything other than `--build-from-source install` (e.g. `brew postinstall mysql` crashes because `buildpath` is null).

I've checked in the updated code using `buildpath` var and moved the logic to the `install` function. Let me know if this is the correct approach?

MikeMcQuaid commented on Formula/mysql.rb:
```diff
@@ -123,13 +123,27 @@ def post_install
+    # Create my.cnf that binds to 127.0.0.1 by default
+    unless (etc/"my.cnf").exist?
+      s = <<-EOS.undent
+      # Default Homebrew MySQL server config
```
Indent each line by two more spaces

npinchot replied:
Fixed :)

MikeMcQuaid commented on Formula/mysql.rb:
```diff
@@ -123,13 +123,27 @@ def post_install
     We've installed your MySQL database without a root password. To secure it run:
         mysql_secure_installation
 
+    We've installed a default server config which only allows connections from localhost.
```
`MySQL is configured to only allow connections from localhost by default` and skip mention of the configuration as it's in the default, typical location.

npinchot replied:
Done :)

MikeMcQuaid commented:
@npinchot This is looking good to me now. If you wanted to be the best person who ever lived: could you make this same change in:
```
mariadb.rb
mariadb@10.0.rb
mysql.rb
mysql@5.5.rb
mysql@5.6.rb
mysql-cluster.rb
percona-server.rb
percona-server@5.5.rb
percona-server@5.6.rb
```

Thanks so much!

npinchot commented:
@MikeMcQuaid Awesome, thanks! I'm happy to help - Homebrew has saved me countless hours, thanks to you and the other folks on this project! I've updated the other files you noted. There was already a different implementation in `mysql-cluster.rb` for installing a default `my.cnf` so I added the `bind-address` there instead of writing another one with `etc.install`. Let me know if you need me to squash all these commits or any other changes to get it merged.

npinchot commented:
Looking into the build failure. Looks like it's with `percona-server@5.6.rb` and `percona-server@5.5.rb`.

npinchot commented:
Looks like here's the issues.

```
percona-server@5.5:
  * macOS has been 64-bit only so build.universal? is deprecated.
  * macOS has been 64-bit only since 10.6 so ENV.universal_binary is deprecated.
Error: 2 problems in 1 formula
percona-server@5.6:
  * macOS has been 64-bit only so build.universal? is deprecated.
  * macOS has been 64-bit only since 10.6 so ENV.universal_binary is deprecated.
Error: 2 problems in 1 formula
```

There was also a problem with a missing `test do` block for `percona-server@5.5` but I fixed by copying from `percona-server@5.6` (I ran `brew install --build-from-source percona-server@5.5` after adding the `test do` block and it worked no issues) . Not sure what to do about the universal errors, if anything? I tried to find the corresponding code in `persona-server` and it looks like that block of code has just been removed. Let me know if we need to fix those or just ignore.

npinchot commented:
@MikeMcQuaid Thanks for fixing up the universal build errors! Let me know if anything else you need me to do on this PR :)

MikeMcQuaid commented:
@npinchot No worries! Should be good to go, just need to finish stuff I'm doing first.

MikeMcQuaid commented:
Thanks for your first contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

Just for future reference we normally prefer only one formula changed per commit.

npinchot commented:
@MikeMcQuaid No problem! Happy to help out :)
I hate to be the bearer of bad news but it looks like the `etc.install "my.cnf"` isn't working when installed from a bottle? I did a `brew cleanup` and a `brew uninstall mysql` and then `brew install mysql` and I don't get a `/usr/local/etc/my.cnf` file. Of course I made sure my brew formulas are all up to date locally/etc. Just to make sure it's nothing funky with my machine after editing formulas I tried the `brew install mysql` on a fresh macOS Sierra VM (after `brew update`) and same problem :(

ilovezfs commented:
@MikeMcQuaid looks like it happened again. No config files in the bottle.

ilovezfs commented:
Checking the El Capitan bottles (I haven't checked Sierra or Yosemite), only mariadb and mysql@5.5 got the config files:
```
iMac-TMP:Homebrew joe$ ls -1tr |grep 'mariadb\|mysql\|percona'|while read f; do echo "";echo $f:; tar -tf $f|grep .bottle; done

mariadb@10.0-10.0.30.el_capitan.bottle.1.tar.gz:

percona-server@5.6-5.6.35-81.0.el_capitan.bottle.1.tar.gz:

percona-server@5.5-5.5.54-38.7.el_capitan.bottle.1.tar.gz:

percona-server-5.7.17-13.el_capitan.bottle.1.tar.gz:

mariadb-10.1.22.el_capitan.bottle.1.tar.gz:
mariadb/10.1.22/.bottle/
mariadb/10.1.22/.bottle/etc/
mariadb/10.1.22/.bottle/etc/init.d/
mariadb/10.1.22/.bottle/etc/logrotate.d/
mariadb/10.1.22/.bottle/etc/my.cnf
mariadb/10.1.22/.bottle/etc/my.cnf.d/
mariadb/10.1.22/.bottle/etc/my.cnf.default
mariadb/10.1.22/.bottle/etc/logrotate.d/mysql
mariadb/10.1.22/.bottle/etc/init.d/mysql

mysql@5.6-5.6.36.el_capitan.bottle.1.tar.gz:

mysql-5.7.18.el_capitan.bottle.1.tar.gz:

mysql@5.5-5.5.56.el_capitan.bottle.tar.gz:
mysql@5.5/5.5.56/.bottle/
mysql@5.5/5.5.56/.bottle/etc/
mysql@5.5/5.5.56/.bottle/etc/my.cnf
```

ilovezfs commented:
OK, so the only reason `mysql@5.5` isn't also broken is that it was upgraded in https://github.com/Homebrew/homebrew-core/pull/13184

That means only `mariadb` got the config file, which makes sense since the `mariadb` bottle was built first by `test-bot`, and the config file wasn't removed before the other bottles were built.
